File: frontend/lib/governance.ts
import { cookies } from "next/headers";
import { cache } from "react";

export type CorrectionPlan = {
  id: number;
  user_id: number;
  user_full_name: string;
  period: string;
  stage: "load_reduction" | "warning" | "termination";
  reason: string;
  load_reduction_percent: number | null;
  status: "active" | "resolved";
  started_by_id: number;
  started_at: string;
  resolved_at: string | null;
};

export type ForceMajeureDeclaration = {
  id: number;
  declared_by_id: number;
  affected_user_id: number | null;
  period: string;
  category: string;
  description: string;
  extension_days: number;
  status: "pending" | "acknowledged" | "resolved";
  started_at: string;
  resolved_at: string | null;
};

export type Incentive = {
  id: number;
  user_id: number;
  user_full_name: string;
  period: string;
  type: "monetary" | "title" | "certificate" | "training_trip" | "other";
  title: string;
  description: string | null;
  amount: number | null;
  decided_by_id: number;
  decided_at: string;
  status: "active" | "revoked";
  revoked_at: string | null;
  revoked_reason: string | null;
};

function internalUrl() {
  return process.env.INTERNAL_API_URL ?? "http://backend:8000";
}

export const getCorrectionPlans = cache(async (locale: string): Promise<CorrectionPlan[]> => {
  const cookieStore = cookies();
  const res = await fetch(`${internalUrl()}/api/v1/correction-plans`, {
    headers: { cookie: cookieStore.toString(), "Accept-Language": locale },
    cache: "no-store",
  });
  if (!res.ok) return [];
  return res.json();
});

export const getForceMajeureDeclarations = cache(async (locale: string): Promise<ForceMajeureDeclaration[]> => {
  const cookieStore = cookies();
  const res = await fetch(`${internalUrl()}/api/v1/force-majeure`, {
    headers: { cookie: cookieStore.toString(), "Accept-Language": locale },
    cache: "no-store",
  });
  if (!res.ok) return [];
  return res.json();
});

export const getIncentives = cache(async (locale: string): Promise<Incentive[]> => {
  const cookieStore = cookies();
  const res = await fetch(`${internalUrl()}/api/v1/incentives`, {
    headers: { cookie: cookieStore.toString(), "Accept-Language": locale },
    cache: "no-store",
  });
  if (!res.ok) return [];
  return res.json();
});

File: backend/app/schemas/governance.py
from datetime import datetime

from pydantic import BaseModel, Field


class CategoryReviewerIn(BaseModel):
    kpi_category_id: int
    user_id: int


class CategoryReviewerOut(BaseModel):
    id: int
    kpi_category_id: int
    category_name: str
    user_id: int
    user_full_name: str


class CategoryHeadApproverIn(BaseModel):
    kpi_category_id: int
    kpi_indicator_id: int | None = None
    user_id: int


class CategoryHeadApproverOut(BaseModel):
    id: int
    kpi_category_id: int
    category_name: str
    kpi_indicator_id: int | None
    indicator_name: str | None
    user_id: int
    user_full_name: str


class CorrectionPlanCreate(BaseModel):
    user_id: int
    period: str = Field(min_length=1)
    stage: str
    reason: str = Field(min_length=1)
    load_reduction_percent: float | None = Field(default=None, ge=0, le=100)


class CorrectionPlanOut(BaseModel):
    id: int
    user_id: int
    user_full_name: str
    period: str
    stage: str
    reason: str
    load_reduction_percent: float | None
    status: str
    started_by_id: int
    started_at: datetime
    resolved_at: datetime | None


class ForceMajeureCreate(BaseModel):
    affected_user_id: int | None = None
    period: str = Field(min_length=1)
    category: str
    description: str = Field(min_length=1)
    extension_days: int = Field(default=0, ge=0)


class ForceMajeureOut(BaseModel):
    id: int
    declared_by_id: int
    affected_user_id: int | None
    period: str
    category: str
    description: str
    extension_days: int
    status: str
    started_at: datetime
    resolved_at: datetime | None


class IncentiveCreate(BaseModel):
    user_id: int
    period: str = Field(min_length=1)
    type: str
    title: str = Field(min_length=1)
    description: str | None = None
    amount: float | None = Field(default=None, ge=0)


class IncentiveRevoke(BaseModel):
    revoked_reason: str = Field(min_length=1)


class IncentiveOut(BaseModel):
    id: int
    user_id: int
    user_full_name: str
    period: str
    type: str
    title: str
    description: str | None
    amount: float | None
    decided_by_id: int
    decided_at: datetime
    status: str
    revoked_at: datetime | None
    revoked_reason: str | None
